adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/AbstractVertxBasedCoapAdapter.java

```diff
@@ -0,0 +1,419 @@
+     * follows:
+     * <ul>
+     * <li>2.04 (Changed) - if the message has been sent to the Hono server.</li>
+     * <li>4.01 (Unauthorized) - if the device could not be authorized.</li>
+     * <li>4.03 (Forbidden) - if the tenant/device is not authorized to send messages.</li>
+     * <li>4.06 (Not Acceptable) - if the message is misformed.</li>
+     * <li>5.00 (Internal Server Error) - if the message could not be sent caused by an unspecific processing
+     * error.</li>
+     * <li>5.03 (Service Unavailable) - if the message could not be sent to the Hono server, e.g. due to lack of
+     * connection or credit.</li>
+     * <li>?.?? (Generic mapped HTPP error) - if the message could not be sent caused by an specific processing error.
+     * See {@link CoapErrorResponse}.</li>
+     * </ul>
+     * 
+     * @param exchange coap message exchange
+     * @param authenticatedDevice authenticated device
+     * @param device message's origin device
+     * @param payload message payload
+     * @param contentType content type of message payload
+     * @param senderTracker hono message sender tracker
+     * @param endpointName message destination endpoint name
+     */
+    private void doUploadMessage(final CoapExchange exchange, final Device authenticatedDevice, final Device device,
+            final Buffer payload, final String contentType, final Future<MessageSender> senderTracker,
+            final String endpointName) {
+
+        if (contentType == null) {
+            exchange.respond(ResponseCode.NOT_ACCEPTABLE);
+        } else if (payload == null || payload.length() == 0) {
+            exchange.respond(ResponseCode.NOT_ACCEPTABLE);
+        } else {
+
+            final Future<JsonObject> tokenTracker = getRegistrationAssertion(device.getTenantId(), device.getDeviceId(),
+                    authenticatedDevice);
+            final Future<TenantObject> tenantConfigTracker = getTenantConfiguration(authenticatedDevice.getTenantId());
+
+            CompositeFuture.all(tokenTracker, senderTracker, tenantConfigTracker).compose(ok -> {
+
+                if (tenantConfigTracker.result().isAdapterEnabled(getTypeName())) {
+                    final MessageSender sender = senderTracker.result();
+                    final Message downstreamMessage = newMessage(
+                            ResourceIdentifier.from(endpointName, authenticatedDevice.getTenantId(),
+                                    device.getDeviceId()),
+                            sender.isRegistrationAssertionRequired(),
+                            "/" + exchange.getRequestOptions().getUriPathString(),
+                            contentType,
+                            payload,
+                            tokenTracker.result(),
+                            null);
+                    customizeDownstreamMessage(downstreamMessage, exchange);
+                    // wait for outcome, ensure message order, if CoAP NSTART-1 is used.
```

Comment:
The specific method to use for sending a telemetry message downstream mainly depends on the QoS that the device wants to use. If its QoS 0 (fire-and-forget, the default), then `sender.send()` is more appropriate because it allows the response to be sent immediately to the device (always 2.04 in CoAP I assume).
If the device wants to use QoS 1, then `sender.sendAndWaitForOutcome` is appropriate because the returned Future will only be completed successfully, once the downstream peer has accepted the message.

For events we only support QoS 1, thus `sender.sendAndWaitForOutcome` is always the right choice.

Reply:
I'm not really sure, what you want to say.

In CoAP the most common pattern, as far as I know, is a CON request. That seems to be closest to QoS 1, or do you have a different opinion? 

NON request seems to be rarely used, so I just didn't pay attention to them.
I'm not sure, what the target should now be, try to make a smallest possible PR?
Or to support rare use-cases?

Reply:
This is not about the QoS of the CoAP request itself. I agree that we should use CON CoAP requests. The question here is, when do we return the CoAP response to the device. When a device uploads telemetry data, we do not make any guarantees as to whether the data is actually processed or gets lost while sending it downstream (this is what we refer to as QoS 0 in Hono). When processing the CoAP request, the adapter therefore doesn't need to wait for the acknowledgment of the downstream AMQP 1.0 peer that it has received the message. This behavior is reflected by the `MessageSender.send()` method. This method immediately returns a succeeded Future, if it was at all possible to send the message downstream, i.e. a connection exists and we have some credits. In constrast, the `MessageSender.sendAndWaitForOutcome()` method sends the message, but completes the returned Future only when the downstream peers acks the reception of the message. We refer to this behavior with QoS 1 in Hono. So, depending on what type of message the device uploads, the adapters invoke *send* (telemetry) or  *sendAndWaitForOutcome* (event) and return a response to the device based on the Future's outcome.

Reply:
??? What's that ???

First you state:
"The specific method to use for sending a telemetry message downstream mainly depends on the QoS that the device wants to use. "

And now:

"So, depending on what type of message the device uploads, the adapters invoke send (telemetry) or sendAndWaitForOutcome (event) and return a response to the device based on the Future's outcome."

If it takes just a few days, to change your mind, would it be required to make it configurable?

Reply:
I have not changed my mind, but I might not have been precise enough about the whole thing in my previous comments.

I think the confusion stems from the fact that a device may send telemetry using either QoS 0 (which is the default) or QoS 1 (which it needs to indicate explicitly, e.g. in the HTTP adapter by including the *qos-level* header or query parameter in the request).

On the other hand, events published by devices are *always* processed with QoS 1. You may want to refer to the other protocol adapters' corresponding *doUploadMessage* methods for how they determine the intended QoS and then use the corresponding methods to send the messages downstream.

Reply:
I'm not sure, why a client should retransmit its message, if the server then drops it :-).
I can add support for NON, if you think that's used (we are anyway above the 1000 lines).
Or just define the way, you want to mark the CoAP request to be either processed as QoS 0 or 1.

Reply:
I am not trying to prescribe any particular way of implementing it. I merely tried to explain the overall idea of the different QoS levels and what their purpose within Hono is. Mapping these concepts to the transport protocol at hand, is the responsibility of the protocol adapter. That's why we are discussing this in the first place. I would welcome any proposals you have for doing so and if I understand correctly, you propose to use the (CoAP) request type to indicate the intended QoS. It sounds reasonable to me to support this, as it wouldn't require transmission of an additional header or query param. However, if I am not mistaken, you also seemed to have pointed out that some clients would never want to use a NON message at all (or may not even support sending them). If so, then it would still make sense to also support a corresponding option or query param, wouldn't it?

So, we would end up with something like this:
* if the request is for the telemetry endpoint and is a NON, use `MessageSender.send()`
* if the request is for the telemetry endpoint, is a CON and includes an option/query param indicating QoS 0,  use `MessageSender.send()` and send the response based on the outcome of the future returned by the send method
* if the request is for the event endpoint and is a CON, use `MessageSender.sendAndWaitForOutcome()` and send the response based on the outcome of the future

Does this make sense from your point of view?

Reply:
I wouldn't make it too complex, without having experience, how it is really used.
So I would got for NON => QoS 0 and CON => QoS 1,
 as long, as someone has no real use case, where something additional is required.

Basically the question behind that seems to be: 
is "use telemetry and the server may drop messages by random" really one major flavour,
or "when messages must be dropped, let the application/client decide which messages are dropped" would be more appreciated.
(My personal forecast is, many will start with "telemetry and random drop" but will end up in "selective drop".)

Reply:
sounds good to me :-)